packages/shadycss/src/shadow-parts.js

```diff
@@ -165,3 +165,47 @@ export function formatShadyPartSelector(
     })
     .join('');
 }
+
+/*  eslint-disable no-unused-vars */
+/**
+ * Perform any needed Shadow Parts updates after a ShadyDOM insertBefore call
+ * has been made.
+ *
+ * @param {!HTMLElement} parentNode
+ * @param {!HTMLElement} newNode
+ * @param {?HTMLElement} referenceNode
+ * @return {void}
+ */
+export function onInsertBefore(parentNode, newNode, referenceNode) {
+  /* eslint-enable no-unused-vars */
+  if (!parentNode.getRootNode) {
+    // TODO(aomarks) Why is it in noPatch mode on Chrome 41 and other older
+    // browsers that getRootNode is sometimes undefined?
+    return;
+  }
+  const root = parentNode.getRootNode();
+  if (root === document) {
+    // Parts in the document scope would never have any effect. Return early so
+    // we don't waste time querying it.
+    return;
+  }
+  const parts = parentNode.querySelectorAll('[part]');
```

Comment:
Also consider doing this in `prepareTemplateDom` so that only dynamic parts need to be changed here.

Reply:
This should be searching `newNode` and `newNode.querySelectorAll`, not `parentNode.querySelectorAll`.

Reply:
You'd need to add support for knowing when to not support the querySelector, but there's precedent for this: https://github.com/webcomponents/polyfills/blob/master/packages/shadydom/src/patches/Node.js#L309.

Reply:
> Does this do a `querySelectorAll` on every call to `insertBefore`? If so, have you measured the performance impact of this? I'd expect it to be noticeable. ShadyDOM already conditionally does a walk during insertBefore (for example, to find slots); maybe consider a tighter integration to be able to participate in that walk.

No benchmarks yet. I will add them in an upcoming task: https://github.com/webcomponents/polyfills/issues/344

Will look into replacing the `querySelectorAll` with an integration into the ShadyDOM walk as an upcoming task: https://github.com/webcomponents/polyfills/issues/345

Reply:
> Also, it seems like this could cut down on the cost of the `querySelectorAll` since I guess I could store some kind of index of the part locations (or at the very least, a bit to tell me if it is even worth querying for parts), but I couldn't actually fully pre-process the template to add the `shady-part` attribute, since the value of that attribute depends on knowing the super-scope, which we don't know until we're connected. Does that sound right?

@sorvell any thoughts on this? I've filed https://github.com/webcomponents/polyfills/issues/346 for future investigation.